### ISSUE TITLE
Learning rate reduction in ReduceLROnPlateau scheduler occurs one epoch later than expected due to patience parameter comparison

### ISSUE THREAD

**Issue by hxfdanger** [pytorch/pytorch#119763]
## 🐛 Describe the bug

The issue lies in the `ReduceLROnPlateau` scheduler's behavior when checking whether to reduce the learning rate based on the `patience` parameter. According to the PyTorch documentation, `patience` is described as the number of epochs with no improvement after which the learning rate will be reduced. For example, if `patience = 2`, then the scheduler should ignore the first 2 epochs with no improvement, and only decrease the learning rate after the 3rd epoch if there is still no improvement. 

However, in the current implementation, with `patience = 2`, the learning rate is actually reduced at the 4th epoch, which contradicts the expected behavior.

```python
if self.num_bad_epochs > self.patience:
    self._reduce_lr(epoch)
    self.cooldown_counter = self.cooldown
    self.num_bad_epochs = 0
```
 
 ## Possible solution

To address this issue, we can modify the condition in the `ReduceLROnPlateau` scheduler's `step` method to ensure that the learning rate reduction occurs after exactly `patience` epochs without improvement. 

```python
if self.num_bad_epochs >= self.patience:
    self._reduce_lr(epoch)
    self.cooldown_counter = self.cooldown
    self.num_bad_epochs = 0
   ```

### Versions

No Version needed. 

cc @svekars @brycebortree @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

**Comment by janeyx99** [comment]
Oh, I see. In this case, we _could_ technically fix the code, but that would be BC breaking, and since the definition of `patience` is not standardized, it would be best to fix the documentation to align with what `patience` actually is.

In this case, `patience` is more like...the number of allowed "bad" epochs with no improvement, after which the lr will be reduced. We will accept a PR with more clarified docs.

**Comment by hxfdanger** [comment]
Thank you for your response. 

Based on the current implementation, if `patience = 2`, then the user would indeed need to wait for 4 epochs before the learning rate is adjusted. Similarly, if `patience = 1`, the user would need to wait for 3 epochs, and if `patience = 0`, they would need to wait for 2 epochs. However, the principle is that the user should wait at least one epoch (to calculate the difference) before the learning rate can evolve. Therefore, the idea is to determine which value of `patience` would ensure that the learning rate evolves at each epoch if there is no improvement in the metric.

With the current implementation, achieving this principle is impossible. Even if the user chooses `patience = -1`, it would still modify the learning rate even if there is no improvement, thus breaking the principle. 

Hence, it's crucial to fix the code to align with this principle. It's not just about whether the definition is standardized or not, but rather ensuring clarity in the principle and implementation to accommodate all use cases.

Therefore, as an ML engineer, I believe that the correction in the code is necessary. I leave it up to you to choose the solution that best suits your needs.

**Comment by hxfdanger** [comment]
The proposed solution suggests a technical clarification regarding the behavior of the `patience` parameter in the `ReduceLROnPlateau` scheduler. Specifically, if `patience` can take the value 0, it implies that no "bad" epochs are tolerated before reducing the learning rate. In this scenario, it would be logical to reset the `num_bad_epochs` counter to -1, indicating that no epochs have yet been evaluated.

However, to maintain consistency and ease of understanding in the implementation, it is advisable to limit the minimum value of `patience` to 1. This ensures that even if the user specifies a minimal patience, the scheduler will still operate according to expectations, reducing the risk of unexpected behavior.

Therefore, the proposed solution establishes clear guidelines for the use of the `patience` parameter while maintaining the robustness and predictability of the `ReduceLROnPlateau` scheduler implementation.

**Comment by janeyx99** [comment]
Ah, let me clarify the current behavior with hypothetical `patience` equal to 0 and 1 in the case where you'd expect the lr to be adjusted/every epoch that can be bad ends up being bad.

Case 1: `patience=0`
**1st epoch**: You run the first epoch to get a baseline `best`. Here, since there was no previous baseline (self.best is init'd to the worst number), the first epoch is good no matter what.
**2nd epoch**: Now you do have a baseline, and this epoch performs worse. `num_bad_epochs` += 1, and 1 > a patience of 0, so, at the end of this step, the lr will be reduced, and `num_bad_epochs` reset to 0.
**3rd epoch**: The lr is reduced, but this epoch still performs worse! `num_bad_epochs` += 1, and 1 > a patience of 0, so, at the end of this step, the lr will be reduced, and `num_bad_epochs` reset to 0.
...
**nth epoch**: The lr is reduced, but this epoch still performs worse! `num_bad_epochs` += 1, and 1 > a patience of 0, so, at the end of this step, the lr will be reduced, and `num_bad_epochs` reset to 0.

Note that in the above, it is true that you waited 2 epochs for the lr to be reduced, but that is reasonable behavior as the first epoch computes the baseline `best`. For every epoch thereafter, since you have "no" patience, the lr gets reduced every epoch. This is expected.

Case 2: `patience=1`
**1st epoch**: You run the first epoch to get a baseline `best`. Here, since there was no previous baseline (self.best is init'd to the worst number), the first epoch is good no matter what.
**2nd epoch**: Now you do have a baseline, and this epoch performs worse. `num_bad_epochs` += 1, and 1 is NOT > a patience of 1.
**3rd epoch**: This epoch keeps performing worse! `num_bad_epochs` += 1, and 2 > a patience of 1, so, at the end of this step, the lr will be reduced, and `num_bad_epochs` reset to 0.
**4th epoch**: The lr is reduced, but this epoch still performs worse! `num_bad_epochs` += 1, and 1 is NOT > a patience of 1.
**5th epoch**: This epoch keeps performing worse! `num_bad_epochs` += 1, and 2 > a patience of 1, so, at the end of this step, the lr will be reduced, and `num_bad_epochs` reset to 0.
...

Now, with a patience of 1, note that there is now a hesitation before lr gets reduced. Compared to patience of 0, where LR gets reduced every epoch, this case has LR being reduced every other epoch.

**Comment by hxfdanger** [comment]
You are completely right !

**Comment by lancerts** [comment]
Just to clarify in the following example, although the lr keep reducing after 2nd epoch, the num_bad_epochs don't reset to 0. It keeps +1 if the epoch keeps performing worse. It only reset to zero if the epoch performs better than the baseline.

> Case 1: patience=0
> 1st epoch: You run the first epoch to get a baseline best. Here, since there was no previous baseline (self.best is init'd to the worst number), the first epoch is good no matter what.
> 2nd epoch: Now you do have a baseline, and this epoch performs worse. num_bad_epochs += 1, and 1 > a patience of 0, so, at the end of this step, the lr will be reduced, and num_bad_epochs reset to 0.
> 3rd epoch: The lr is reduced, but this epoch still performs worse! num_bad_epochs += 1, and 1 > a patience of 0, so, at the end of this step, the lr will be reduced, and num_bad_epochs reset to 0.
> ...
> nth epoch: The lr is reduced, but this epoch still performs worse! num_bad_epochs += 1, and 1 > a patience of 0, so, at the end of this step, the lr will be reduced, and num_bad_epochs reset to 0.
